vochain/state_snapshot.go

```
@@ -0,0 +1,416 @@
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	b, err := io.ReadAll(s.currentTreeReadBuffer)
+	if err != nil {
+		return nil, err
+	}
+	if len(b) == 0 {
+		return nil, io.EOF
+	}
+	return b, nil
+}
+
+// Header returns the header for the snapshot containing the information
+// about all the merkle trees.
+func (s *StateSnapshot) Header() *SnapshotHeader {
+	return &s.header
+}
+
+// TreeHeader returns the header for the current tree.
+func (s *StateSnapshot) TreeHeader() *SnapshotHeaderTree {
+	return &s.header.Trees[s.currentTree]
+}
+
+// Path returns the file path of the snapshot file currently used.
+func (s *StateSnapshot) Path() string {
+	return s.path
+}
+
+// Create starts the creation of a new snapshot as a disk file.
+// This method must be called only once and its operation is oposed to `Open`.
+func (s *StateSnapshot) Create(filePath string) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	var err error
+	s.header.Version = snapshotHeaderVersion
+	s.file, err = os.Create(filePath + ".tmp")
+	s.path = filePath
+	return err
+}
+
+// AddTree adds a new tree to the snapshot. `Create` needs to be called first.
+func (s *StateSnapshot) AddTree(name, parent string, root []byte) {
+	s.lock.Lock() // only 1 tree at time is allowed
+	s.header.Trees = append(s.header.Trees, SnapshotHeaderTree{
+		Name:   name,
+		Parent: parent,
+		Root:   root,
+		Size:   0,
+	})
+}
+
+// EndTree finishes the addition of a tree. This method should be called after `AddTree`.
+func (s *StateSnapshot) EndTree() {
+	s.currentTree++
+	s.lock.Unlock()
+}
+
+// Save builds the snapshot started with `Create` and stores in disk its contents.
+// After calling this method the snapshot is finished.
+// `EndTree` must be called before saving.
+func (s *StateSnapshot) Save() error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	// create the final file
+	finalFile, err := os.Create(s.path)
+	defer func() {
+		finalFile.Close()
+		if err != nil {
+			log.Error(err)
+		}
```

this is weird. you're not actually checking the close error? also, you should return the error, not log it. so this should go at the end of the function instead, and here you should do a `defer finalFile.Close()`.

---

right